semgrep-core/src/tainting/Taint_lval_env.ml

```diff
@@ -69,36 +69,50 @@ let union le1 le2 =
  * not represent the exact same object as the original l-value, but an
  * overapproximation. For example, the normalized l-value of `x[i]` will be `x`,
  * so the taints of any element of an array are tracked via the array itself. *)
-let rec normalize_lval lval =
+let normalize_lval lval =
+  let open Common in
   let { IL.base; rev_offset } = lval in
-  match lval.IL.rev_offset with
-  | [] -> (
-      (* Base case, no offset; we can only track variables. *)
-      match base with
-      | Var _ -> Some lval
-      | VarSpecial _
-      | Mem _ ->
-          None)
-  | _ :: rev_offset' -> (
-      (* Must find the longest prefix of the form x.a_1. ... . a_N. *)
-      let is_dots = LV.is_dots_offset rev_offset in
-      match base with
-      | Var _ when is_dots -> Some lval
-      | VarSpecial _ when is_dots -> (
-          (* this.x.a_1. ... . a_N becomes x.a_1. ... . a_N *)
-          match List.rev lval.rev_offset with
-          | { o = IL.Dot var; _ } :: offset' ->
-              Some { IL.base = Var var; rev_offset = List.rev offset' }
-          | []
-          | { o = IL.Index _; _ } :: _ ->
-              logger#error "normalize_lval: Impossible happened";
-              None)
-      | Var _
-      | VarSpecial _ ->
-          (* Offset-chain is not of the form .a_1. ... . a_N, so drop the last
-           * offset until the condition is met or we reach the base case. *)
-          normalize_lval { base; rev_offset = rev_offset' }
-      | Mem _ -> None)
+  let* base, rev_offset =
+    match base with
+    | Var _ -> Some (base, rev_offset)
+    | VarSpecial _ -> (
+        (* this.x.a_1. ... . a_N becomes x.a_1. ... . a_N *)
+        match List.rev lval.rev_offset with
+        | { o = IL.Dot var; _ } :: offset' -> Some (Var var, List.rev offset')
+        | [] -> None
+        | { o = IL.Index _; _ } :: _ ->
+            logger#error "normalize_lval: Impossible happened";
+            None)
+    | Mem _ -> None
+  in
+  let rev_offset =
+    rev_offset
+    |> List.filter (fun o ->
+           match o.IL.o with
+           | IL.Dot _ -> true
+           | IL.Index _ -> false)
```

Comment:
To confirm my understanding, is it accurate to say here `TODO make tainting through arrays index-sensitive`?

Reply:
Yes! We would need to check here for constant indexes, but there is also some extra work to do when looking up the lvalues. Since you may taint `x.a` and then look for `x.a[1]` and it should tell you it's tainted.